neuraxle/data_container.py

```diff
@@ -0,0 +1,242 @@
+        """
+        Hash :class:`DataContainer`.current_ids, data inputs, and hyperparameters together into one id.
+
+        :return: single hashed current id for all of the current ids
+        :rtype: str
+        """
+        m = hashlib.md5()
+        for current_id in self.current_ids:
+            m.update(str.encode(current_id))
+        return m.hexdigest()
+
+    def convolved_1d(self, stride, kernel_size) -> Iterable['DataContainer']:
+        """
+        Returns an iterator that iterates through batches of the DataContainer.
+
+        :param stride: step size for the convolution operation
+        :param kernel_size:
+        :return: an iterator of DataContainer
+        :rtype: Iterable[DataContainer]
+
+        .. seealso::
+            `<https://github.com/guillaume-chevalier/python-conv-lib>`_
+        """
+        conv_current_ids = convolved_1d(stride=stride, iterable=self.current_ids, kernel_size=kernel_size)
+        conv_data_inputs = convolved_1d(stride=stride, iterable=self.data_inputs, kernel_size=kernel_size)
+        conv_expected_outputs = convolved_1d(stride=stride, iterable=self.expected_outputs, kernel_size=kernel_size)
+
+        for current_ids, data_inputs, expected_outputs in zip(conv_current_ids, conv_data_inputs,
+                                                              conv_expected_outputs):
+            yield DataContainer(
+                current_ids=current_ids,
+                data_inputs=data_inputs,
+                expected_outputs=expected_outputs
+            )
+
+    def __iter__(self):
+        """
+        Iter method returns a zip of all of the current_ids, data_inputs, and expected_outputs in the data container.
+
+        :return: iterator of tuples containing current_ids, data_inputs, and expected outputs
+        :rtype: Iterator[Tuple]
+        """
+        current_ids = self.current_ids
+        if self.current_ids is None:
+            current_ids = [None] * len(self.data_inputs)
+
+        expected_outputs = self.expected_outputs
+        if self.expected_outputs is None:
+            expected_outputs = [None] * len(self.data_inputs)
+
+        return zip(current_ids, self.data_inputs, expected_outputs)
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        return self.__class__.__name__ + "(current_ids=" + repr(list(self.current_ids)) + ", ...)"
+
+    def __len__(self):
+        return len(self.data_inputs)
+
+
+class ExpandedDataContainer(DataContainer):
+    """
+    Sub class of DataContainer to expand data container dimension.
+
+    .. seealso::
+        :class:`ExpandedDataContainer`,
+    """
+    def __init__(self, current_ids, data_inputs, expected_outputs, old_current_ids):
+        DataContainer.__init__(
+            self,
+            current_ids,
+            data_inputs,
+            expected_outputs
+        )
+
+        self.old_current_ids = old_current_ids
+
+    def reduce_dim(self) -> 'DataContainer':
```

Comment:
I know there is a cool mathematical name for this but I forgot it. It looks like [np.squeeze](https://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.squeeze.html) but it's not squeeze exactly. Looks fine anyways.